app/scripts/controllers/permissions/permissionsLog.js

```diff
@@ -0,0 +1,306 @@
+      ) {
+
+        activityEntry = this.logActivity(req, isInternal)
+
+        if (method === `${WALLET_PREFIX}requestPermissions`) {
+          // get the corresponding methods from the requested permissions
+          requestedMethods = this.getRequestedMethods(req)
+        }
+      } else if (method === 'eth_requestAccounts') {
+
+        // eth_requestAccounts is a special case; we need to extract the accounts
+        // from it
+        activityEntry = this.logActivity(req, isInternal)
+        requestedMethods = [ 'eth_accounts' ]
+      } else {
+        // no-op
+        return next()
+      }
+
+      // call next with a return handler for capturing the response
+      next(cb => {
+
+        const time = Date.now()
+        this.addResponse(activityEntry, res, time)
+
+        if (!res.error && requestedMethods) {
+          // any permissions or accounts changes will be recorded on the response,
+          // so we only log permissions history here
+          this.logPermissionsHistory(
+            requestedMethods, origin, res.result, time,
+            method === 'eth_requestAccounts',
+          )
+        }
+        cb()
+      })
+    }
+  }
+
+  // creates and commits an activity log entry, without response data
+  logActivity (request, isInternal) {
+    const activityEntry = {
+      id: request.id,
+      method: request.method,
+      methodType: isInternal ? 'internal' : 'restricted',
+      origin: request.origin,
+      request: cloneObj(request),
+      requestTime: Date.now(),
+      response: null,
+      responseTime: null,
+      success: null,
+    }
+    this.commitActivity(activityEntry)
+    return activityEntry
+  }
+
+  // adds response data to an activity entry
+  addResponse (activityEntry, response, time) {
+    if (!response) {
+      return
+    }
+    activityEntry.response = cloneObj(response)
+    activityEntry.responseTime = time
+    activityEntry.success = !response.error
+  }
+
+  // get requested methods from a permissions request
+  getRequestedMethods (request) {
+    if (
+      !request.params ||
+      typeof request.params[0] !== 'object' ||
+      Array.isArray(request.params[0])
+    ) {
+      return null
+    }
+    return Object.keys(request.params[0])
+  }
+
+
+  getAccountsFromPermission (perm) {
+
+    if (perm.parentCapability !== 'eth_accounts' || !perm.caveats) {
+      return []
+    }
+
+    const accounts = {}
+    for (const c of perm.caveats) {
+      if (c.name === CAVEAT_NAMES.exposedAccounts && Array.isArray(c.value)) {
+        for (const v of c.value) {
+          if (isValidAddress(v)) {
+            accounts[v] = true
+          }
+        }
+      }
+    }
+    return Object.keys(accounts)
+  }
+
+  // commit an activity entry to the log
+  commitActivity (entry) {
+
+    const logs = this.getLogStore()
+
+    // add new entry to end of log
+    logs.push(entry)
+
+    // remove oldest log if exceeding size limit
+    if (logs.length > LOG_LIMIT) {
```

Comment:
Previously, this check was `logs.length > LOG_LIMIT - 2`, why was that?

Reply:
Great question! As far as I can tell, no good reason at all.

Reply:
Well once reason for the difference is that it checked _before_ the new log was added, rather than _after_. But that's still one off :thinking: 
It seems the "real" limit previously was one below the log limit. I think this is more correct.